### DISCUSSION TITLE
how to simplify a*cos(x)+b*sin(x)

### DISCUSSION THREAD

**Discussion by Matth32** [sympy/sympy#21489] [Q&A]
Hello,
I would like to simplify a trigonometric function like a*cos(x)+b*sin(x) into Rcos(x-c). 
My expression is:
![image](https://user-images.githubusercontent.com/83581603/118766962-67fd2c00-b87d-11eb-95e3-6c29395eda76.png)
I have tried many functions, I thought that TR10i function of fu library could work but maybe it only works with numbers.

Does anybody have a suggestion?
Thanks

**Comment by Matth32** [comment]
Hello,

Nobody have an idea?

Thanks

**Reply by oscarbenjamin** [reply]
Do you know that a simpler form exists?

**Comment by Matth32** [comment]
Yes, I would like to simplify in this form (made by hand):
![image](https://user-images.githubusercontent.com/83581603/120613857-e8638580-c456-11eb-9d23-3dbe23fd3152.png)

**Reply by Davide-sd** [reply]
Would you be able to provide the hand-written procedure?

**Reply by Matth32** [reply]
Sorry for the delay, here it is:
![image](https://user-images.githubusercontent.com/83581603/128492614-d69cdf25-8c52-4148-b979-7790b95c206c.png)
![image](https://user-images.githubusercontent.com/83581603/128492569-97229697-8756-47db-818b-bab167629174.png)

Thanks

**Reply by oscarbenjamin** [reply]
It would be good to have a function that can combine sin/cos into a single trig function. I wonder what the conditions are for this transformation to be correct. Inverting `tan(theta) = beta/alpha` is not always valid but maybe it is okay when done inside the `cos`: a shift of `pi` would introduce a minus sign but there is a choice of sign for `sqrt(alpha**2 + beta**2)` as well.

**Reply by Matth32** [reply]
I think you are right, atan returns an angle between –pi/2 and pi/2 so it only describes the right side of the trigonometric circle which means positive cosine.
So if the cosine was initially negative (alpha < 0), using atan (b/a) introduces an error of pi which leads to incorrect sign of the result.

For example:
-cos(x)+sin(x)

Applying the formula would give:
√2*cos⁡(x-atan⁡(-1) )=√2*cos⁡(x+π/4)  

Expending back using cosine properties:
√2*(cos⁡(x)*cos⁡(π/4)-sin⁡(x)*sin⁡(π/4))=√2*(cos⁡(x)*√2/2-sin⁡(x)*√2/2)=cos(x)-sin(x)

There is probably a better way to find the angle knowing its sine and cosine, but I don’t have it.
The case where alpha = 0 is also problematic for this factorisation, but there is no need to do it in that case.

**Reply by oscarbenjamin** [reply]
> There is probably a better way to find the angle knowing its sine and cosine, but I don’t have it.

That's `atan2` which SymPy has:
```python
In [11]: [atan2(y, x) for x, y in [(1, 1), (-1, 1), (-1, -1), (1, -1)]]
Out[11]: 
⎡π  3⋅π  -3⋅π   -π ⎤
⎢─, ───, ─────, ───⎥
⎣4   4     4     4 ⎦
```

> The case where alpha = 0 is also problematic for this factorisation, but there is no need to do it in that case.

The `atan2` function handles all of the special cases:
```python
In [12]: [atan2(y, x) for x, y in [(0, 1), (0, -1), (1, 0), (-1, 0)]]
Out[12]: 
⎡π  -π       ⎤
⎢─, ───, 0, π⎥
⎣2   2       ⎦
```

**Reply by Matth32** [reply]
Yes atan2 seems to make the job for that. 
So, if my understanding is right, in the actual code sympy is not able to make the complete simplification and it will be difficult to code a simplification which take all case into account?

**Reply by oscarbenjamin** [reply]
I think it can be done. There just is not a function for this particular rearrangement in SymPy yet (someone needs to make that function).

**Reply by Matth32** [reply]
Ok so is there a kind of pool of requests or something like this? Because I don't have the capability to code this function myself and I think could be an enhancement for sympy.

**Comment by ThePauliPrinciple** [comment]
To answer your question on how to rewrite a*sin(x)+b*cos(x) into R*cos(x-c), we can do this as follows:
```python
import sympy as sym

a,b,c,R,x= sym.symbols('a b c R x')

lhs=a*sym.cos(x)+b*sym.sin(x) # form we have
rhs=R*sym.cos(x-c) # form we want
eq=lhs-rhs #solve expects an equation of the type lhs-rhs=0 instead of lhs=rhs

#solve for the unknowns in the form we want
s1,s2=sym.solve(eq,c)  #as it turns out, two solutions are found for c (makes sense to me in a cos)
s3=sym.solve(eq,R)[0]  #solve returns an array, even if one solution is found```

```
![afbeelding](https://user-images.githubusercontent.com/35560762/135895017-1c23ca1d-d6b0-4bef-a6a4-ed494b7c8093.png)

I do realise this does not obtain the form of R and c you were looking for, but perhaps it is still useful?

**Reply by oscarbenjamin** [reply]
The formula needed is straight-forward:
```python
In [48]: A, B, t = symbols('A, B, t')                                                                

In [49]: expr = sqrt(A**2 + B**2)*cos(t - atan2(B, A))                                               

In [50]: expr                                                                                        
Out[50]: 
   _________                     
  ╱  2    2                      
╲╱  A  + B  ⋅cos(t - atan2(B, A))

In [51]: expr.expand(trig=True)                                                                      
Out[51]: A⋅cos(t) + B⋅sin(t)
```
There just needs to be a function that will:

1. Identify the `cos` and `sin` terms that have the same argument.
2. Extract the coefficients `A` and `B`.
